design/notifications.md

```diff
@@ -0,0 +1,360 @@
+
+# Notifications
+This document outlines the design for notifications in Nessie with an application for Webhooks. It outlines the different
+events that we support, the way the notifications are called, how they are stored and what is the API used for creating,
+listing, updating and deleting notifications for a nessie server.
+
+## Event types
+This section lists out all events we support. All events in Nessie contain two pieces of information: the event time and
+the event type. The event time is the time at which the event was fired, it corresponds closely to when the event
+happened, but it can lag behind a few nanoseconds. The event type is an `UPPER_CASE` string representing the type of the
+event, as an example, commit events have a `"COMMIT"` type.
+
+For the following section, we will use the same schema as Nessie's API for referring to references, branches and tags,
+as well, as other Nessie API model objects, such as CommitMeta and Operation. 
+
+We will refer to them as their type in the Nessie OpenAPI spec, using the `$ref` keyword. 
+
+### Commits
+Commit events are triggered whenever there's a commit to a branch, i.e. whenever the `/trees/branch/{branchName}/commit`
+API in Nessie is called. 
+
+#### Schema
+```yaml
+CommitEvent:
+  type: object
+  properties:
+    type:
+      type: string
+      description: Always COMMIT
+    eventTime:
+      type: string
+      format: date
+      description: The time the event is fired in ISO date format.
+    branch:
+      $ref: Branch
+      description: The branch committed against.
+    newHash:
+      type : string
+      description: The hash of this commit.
+    metadata:
+      $ref: CommitMeta
+      description: This commit's metadata. Containing author, committer, etc.
+    operations:
+      type: array
+      items:
+        $ref: 'Operation'
+```
+
+#### Example
+```json
+{
+  "type" : "COMMIT",
+  "eventTime": "2009-02-14T18:09:10Z",
+  "reference": {
+    "type" : "BRANCH",
+    "name" : "main",
+    "hash" : "6a3efd4f40e88a36"
+  },
+  "metadata": {
+    "author" : "Bob",
+    "message" : "Delete tables.bobby",
+    "commitTime" : "2015-04-14T22:45:00Z"
+  },
+  "newHash" : "e2dad0c9178ef5ff",
+  "operations": [
+    {"type":"DELETE","key":{"elements":["tables","bobby"]}}
+  ]
+}
+```
+
+### Merges
+Merge events are triggered whenever we merge a reference into a branch. In other words, whenever the 
+`/trees/branch/{branchName}/merge` API in Nessie is called.
+
+#### Schema
+```yaml
+MergeEvent:
+  type: object
+  properties:
+    type:
+      type: string
+      description: Always "MERGE"
+    eventTime:
+      type: string
+      format: date
+      description: The time the event was fired in ISO date format.
+    fromRefName:
```

Comment:
`fromRefName` + `fromHash` can be replaced w/ `fromRef` using the `Reference` type, so the reference type is also clear here (branch/tag/detached).